diesel/src/query_dsl/mod.rs

```diff
@@ -836,8 +836,8 @@ pub trait QueryDsl: Sized {
 
     /// Sets the `group by` clause of a query.
     ///
-    /// **Note:** Queries having a `group by` clause require a custom select clause.
-    /// Use `QueryDsl::select()` to specify one
+    /// **Note:** Queries having a `group by` clause require a custom select clause _after_ it.
```

Comment:
Sorry for being pedantic, but that's just not true. The real think is a bit more complex than this simplified statement. 
Each time you call `.select()` it checks that the select clause is valid for the current `.group_by()` clause, so technically speaking it's possible to write queries that set the select clause before calling `.group_by()` as long as the select clause is valid for both, no group by clause and the group by clause you set afterwards.
The simplest example for such a query is the following:
```rust
some_table::table.select(1.into_sql::<Integer>()).group_by(some_table::name)
```
as all literal expressions are valid regardless for the group by clause.

Reply:
That's what I tried, I just swap the position and it does not work. Maybe you can take a look at what I was doing https://github.com/pickfire/diesel-group-by

Reply:
> Please stop to make conclusions about how the whole system works from a single example, especially if the author tells you that your assumptions are not correct.

I understand, but I wanted to try to understand more about it so I did a few more examples, does that mean every example that I do is incorrect?

>  For your example that means user_id, post_id, count(id) would be required to be valid for no group by clause. As this mixes aggregate and non aggregate expressions that's not the case. It is important to note that this is really depended on the expressions used as select clause so we cannot generally say that you need to call select after group_by.

How does the user know in what cases what behavior are there and in what cases what behavior are there? Do they need to experience themselves not being able to compile or we could put some note up front to tell them about their behavior? Maybe the behavior is just too complicated to be documented?

Reply:
> I understand, but I wanted to try to understand more about it so I did a few more examples, does that mean every example that I do is incorrect?

That's all fine. I only ask that you don't try to extrapolate from one example the whole behaviour of a complex generic function complex. I have the feeling that this happened here and that this also happened in #2631. I mean our documentation needs to cover all edge cases, not only that one that you personally hit. I personally have the opinion that it's better to have documentation not mentioning some things, than have documentation mentioning things as impossible that are in fact possible.

> How does the user know in what cases what behavior are there and in what cases what behavior are there? Do they need to experience themselves not being able to compile or we could put some note up front to tell them about their behavior? Maybe the behavior is just too complicated to be documented?

So what are your actual suggestions here? Do you want to see the API changed? Have you concrete proposals for this beside "the current way is bad"? In my opinion this is nothing that can "just" be fixed by adding some half true statement to some random location in our documentation.
The main point here is that SQL itself is complex, especially the interactions between different features. Now diesel chooses to provide a lot guarantees at compile time, that means most of this is already documented as part of the type system.